cli/src/formatted_diagnostic.rs

```diff
@@ -0,0 +1,156 @@
+use regex::Regex;
+use yansi::Paint;
+
+use context::source::Source;
+
+use crate::fix_ariadne_report;
+
+pub fn render_diagnostic<'a>(
+    diagnostic: analyzer::diagnostic::Diagnostic,
+    source_supplier: impl Fn(SourceId) -> Source<'a>,
+) -> io::Result<String> {
+    let mut colors = ColorGenerator::new();
+    let mut colormap = HashMap::new();
+    let mut sources = HashMap::new();
+
+    fn get_color(
+        colors: &mut ColorGenerator,
+        colormap: &mut HashMap<u8, Color>,
+        tag: Option<ObservationTag>,
+    ) -> Color {
+        match tag {
+            Some(ObservationTag::InFault) => Color::Red,
+            Some(ObservationTag::Declaration) => Color::Yellow,
+            Some(ObservationTag::Expected) => Color::Green,
+            Some(ObservationTag::Other(tag)) => {
+                *colormap.entry(tag).or_insert_with(|| colors.next())
+            }
+            None => colors.next(),
+        }
+    }
+
+    let main_source = source_supplier(diagnostic.source);
+    sources.insert(diagnostic.source, main_source);
+
+    let (code, color) = if diagnostic.identifier.critical() {
+        (
+            format!("error[E{:04}]", diagnostic.identifier.code()),
+            Color::Red,
+        )
+    } else {
+        (
+            format!("warn[W{:04}]", diagnostic.identifier.code()),
+            Color::Yellow,
+        )
+    };
+
+    let labels = diagnostic.observations.into_iter().map(|o| {
+        let source = o.foreign_env.unwrap_or(diagnostic.source);
+        let source = *sources
+            .entry(source)
+            .or_insert_with(|| source_supplier(source));
+
+        let mut label = Label::new((source.name, o.segment));
+
+        if let Some(ObservationTag::Expected) = o.tag {
+            label = label.with_order(100)
+        }
+
+        let color = get_color(&mut colors, &mut colormap, o.tag);
+        label = label.with_color(color);
+
+        if let Some(help) = o.help {
+            label = label.with_message(colorize_message(help, color))
+        }
+        label
+    });
+
+    let help = agg_strings(diagnostic.helps);
+    let note = agg_strings(diagnostic.notes);
+
+    let mut builder = Report::build(ReportKind::Custom(&code, color), main_source.name, 0)
+        .with_config(Config::default().with_underlines(false))
+        .with_message(colorize_message(diagnostic.global_message, Color::Blue))
+        .with_labels(labels);
+
+    if let Some(help) = help {
+        builder = builder.with_help(colorize_message(help, Color::Green))
+    }
+
+    if let Some(note) = note {
+        builder = builder.with_note(colorize_message(note, Color::Cyan))
+    }
+
+    let cache = sources
+        .into_values()
+        .map(|s| (s.name, ariadne::Source::from(s.source)))
+        .collect::<HashMap<_, _>>();
+
+    let cache = CacheMap { cache };
+
+    let mut buf = Vec::new();
+    builder.finish().write_for_stdout(cache, &mut buf)?;
```

Comment:
```suggestion
    builder.finish().write(cache, &mut buf)?;
```
The diagnotic will be written to `stderr`